Document builtin entities

Wuzzy2 opened the pull request:
This PR adds a new text file `doc/builtin_entities.txt` which explains how the two builtin entities `__builtin:item` and `__builtin:falling_node` work.

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+
+## Falling node (`__builtin:falling_node`)
+
+Is created when a node with the special group `falling_node=1` starts to fall.
```
Strictly spoken nodes can't fall, hence it might be misleading to mention that `a node [..] starts to fall`.

How about:
```
This entity is created by `minetest.check_for_falling` in place of a node
with the special group `falling_node=1`.
```

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+
+Spawning a falling node with `/spawnentity` won't work. 
+
+Behaviour:
```
`Default behaviour:`

This would also indicate that mods can manipulate this entity. For example: altered horizontal/vertical speed.

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+* It collides with physical objects and `walkable` nodes
+* If the node group `float=1` is set, it also collides with liquid nodes
+* When it hits a solid (=`walkable`) node, it will try to place itself as a
+  node, replacing the node above
```
The following list is not a condition for this statement to apply. They are not related, hence terminating the sentence here clarifies that.
```suggestion
  node, replacing the node above.
```

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+
+### Rendering / supported nodes
+
+Falling nodes try their best to look as close as possible to the original node.
```
`try their best` sounds cute in some way. However, I'd expect different from a documentation file.

```suggestion
Falling nodes have visuals close as possible to the original node.
```

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+
+Other drawtypes still kinda work, but they might look weird.
+
+Supported `paramtype2`s:
```
```suggestion
Supported `paramtype2` values+:
```

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+item, when a node drops as an item and on various other events.
+
+It can be spawned manually with `minetest.add_item`. Spawning this with
+`/spawnentity` won't work.
```
What do you think about converting this to a list?
```
Common cases that spawn a dropped item:

 * Dropped by player
 * The root node of a `attached_node = 1` is removed
 * `minetest.add_item` is called

Spawning this entity using `/spawnentity` will not work.
```

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+* `set_item(self, item)`: A function. Call this to set the item that this
+  entity represents. `item` is an ItemStack.
+* `age`: Age in seconds. When this exceeds `item_entity_ttl`, the entity will
+  die (unless item timeout is disabled)
```
`the entity is removed`

`die` is caused by `set_hp`.

SmallJoker commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+
+### Entity fields
+
+* `set_item(self, item)`: A function. Call this to set the item that this
```
```
Function to initialize this entity to its representing `item` (type `ItemStack`).
```

sfan5 commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+Behaviour:
+
+* Falls down in a straight line (gravity = `movement_gravity` setting)
+* It collides with physical objects and `walkable` nodes
```
At the time you wrote this, "collides with physical objects" was incorrect, but it now is correct.
This should also mentions that it falls through players.

sfan5 replied:
Still relevant:
> This should also mention that it falls through players.

sfan5 commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+* Falls down in a straight line (gravity = `movement_gravity` setting)
+* It collides with physical objects and `walkable` nodes
+* If the node group `float=1` is set, it also collides with liquid nodes
+* When it hits a solid (=`walkable`) node, it will try to place itself as a
```
Missing description of leveled nodes.

sfan5 commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+## Dropped item stack (`__builtin:item`)
+
+This is an item stack in a collectable form. It appears when a player drops an
+item, when a node drops as an item and on various other events.
```
Why "various other events"?
The engine only has player drops item, attached nodes and `add_item`.

sfan5 commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,96 @@
+  (unless item timeout is disabled)
+* It will slide on `slippery` nodes
+* It is subject to gravity (uses `movement_gravity` setting)
+* It collides with physical objects and `walkable` nodes
```
"collides with physical objects"
wrong

Wuzzy2 commented:
Serious question (not meant as an insult or anything): Why don't you do the edits yourself? You seem to best know what you want. You also seem to know much better about the Minetest internals than I do …
The checkbox “allow edits by maintainers” is activated.

SmallJoker commented:
@Wuzzy2 Because I'd prefer to hear your opinion instead of changing relevant parts of what you've written.

Wuzzy2 commented:
> At the time you wrote this, "collides with physical objects" was incorrect, but it now is correct.

> "collides with physical objects"
> wrong

You're sending mixed messages here. :-(

Apart from that, I agree with everything.

sfan5 commented:
> You're sending mixed messages here. :-(

???
Those comments are not on the same line.

Wuzzy2 commented:
I'm confused, and don't know how to proceed. :-(

sfan5 commented:
* `__builtin:falling_node` collides with objects, this is correct in the document
* `__builtin:item` does not collide with objects, this is incorrect in the document

Wuzzy2 commented:
> Missing description of leveled nodes.

What about it?

Also: I have tested collisions in -dev recently.

What I have found:

* Falling nodes do NOT collide with objects/players (so I can't reproduce your claim)
* Dropped items do NOT collide with objects/players (OK)
* Both falling nodes and dropped items are physical (they collide with walkable nodes)

Wuzzy2 commented:
UPDATE!
I (hopefully) fixed all the collision info.
I added description of how leveled nodes are supposed to behave.

I hope everything's resolved now.

Wuzzy2 commented:
Bump after 2 months.

paramat commented on doc/builtin_entities.txt:
```diff
@@ -0,0 +1,101 @@
+
+* Falls down in a straight line (gravity = `movement_gravity` setting)
+* Collides with `walkable` node
+* Collides with all physical objects but players
```
'Collides with all physical objects except players'

rubenwardy replied:
```suggestion
* Collides with all physical objects except players
```

paramat commented:
I will approve and merge this with my requested change.

rubenwardy reviewed (commented):
:+1: if the comment by paramat is addressed
